snapcraft/parts/plugins/matter_sdk_plugin.py

```diff
@@ -0,0 +1,180 @@
+#
+# Copyright 2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The matter SDK plugin."""
+import os
+from typing import Any, Dict, List, Set, cast
+
+from craft_parts import infos, plugins
+from overrides import overrides
+
+# The repository where the matter SDK resides.
+MATTER_SDK_REPO = "https://github.com/project-chip/connectedhomeip"
+
+
+class MatterSdkPluginProperties(plugins.PluginProperties, plugins.PluginModel):
+    """The part properties used by the matter SDK plugin."""
+
+    matter_sdk_version: str
+
+    @classmethod
+    @overrides
+    def unmarshal(cls, data: Dict[str, Any]) -> "MatterSdkPluginProperties":
+        """Populate class attributes from the part specification.
+
+        :param data: A dictionary containing part properties.
+
+        :return: The populated plugin properties data object.
+
+        :raise pydantic.ValidationError: If validation fails.
+        """
+        plugin_data = plugins.extract_plugin_properties(
+            data,
+            plugin_name="matter-sdk",
+            required=["matter_sdk_version"],
+        )
+        return cls(**plugin_data)
+
+
+class MatterSdkPlugin(plugins.Plugin):
+    """A plugin for matter SDK project.
+
+    This plugin uses the common plugin keywords.
+    For more information check the 'plugins' topic.
+
+    Additionally, this plugin uses the following plugin-specific keywords:
+        - matter-sdk-version
+          (str, no default)
+          The matter SDK version to use for the build.
+    """
+
+    properties_class = MatterSdkPluginProperties
+
+    def __init__(
+        self,
+        *,
+        properties: plugins.PluginProperties,
+        part_info: infos.PartInfo,
+    ) -> None:
+        super().__init__(properties=properties, part_info=part_info)
+
+        self.matter_sdk_dir = part_info.part_build_dir
+        self.snap_arch = os.getenv("SNAP_ARCH")
+
+    @overrides
+    def get_pull_commands(self) -> List[str]:
+        options = cast(MatterSdkPluginProperties, self._options)
+        commands = []
+
+        # Clone Matter SDK repository
+        commands.extend(
+            [
+                "if [ ! -d matter ]; then",
+                "    git init",
+                f"   git remote add origin {MATTER_SDK_REPO}",
+                f"   git fetch --depth 1 origin {options.matter_sdk_version}",
+                "    git checkout FETCH_HEAD",
+                "fi",
```

Comment:
Thanks for pointing that out! The matter directory check was leftover from previous changes. There are no scenarios in which the consumer snap can rerun this plugin with a cached matter directory. This is because when consumers override the part that uses this plugin, the plugin directory will be completely removed and rerun, without having a chance to use the existing matter directory. I will remove this `if [ ! -d matter ]` check as it is not necessary.

The reason for not creating a matter directory is that it will make the path less readable. For example, with the current approach, the consumer snap will source the environment variables like [this](https://github.com/canonical/snapcraft/blob/12e1e6ac3e92ddaed064e8d0f07b1894cde6215a/tests/spread/plugins/craft-parts/matter-sdk/snapcraft.yaml#L37):
```
source ../../matter-sdk/build/matter-sdk-env.sh
```

With the matter directory, the path would look like this:
```
source ../../matter-sdk/build/matter/matter-sdk-env.sh
```

Reply:
As the SDK upstream currently has a specific [commit](https://github.com/canonical/snapcraft/blob/12e1e6ac3e92ddaed064e8d0f07b1894cde6215a/tests/spread/plugins/craft-parts/matter-sdk/snapcraft.yaml#L30) with a fix we need, the repo has been [shallow cloned with this commit at depth 1](https://stackoverflow.com/a/43136160/17842525). Once the SDK upstream has tagged this fix in a branch, it could simply be done using git clone and in one line as you suggested.